frontend/src/routes/password.change.tsx

```diff
@@ -0,0 +1,293 @@
+  async loader({ context, abortController: { signal } }) {
+    const viewer = await context.client.query(
+      CURRENT_VIEWER_QUERY,
+      {},
+      { fetchOptions: { signal } },
+    );
+    if (viewer.error) throw viewer.error;
+    if (viewer.data?.viewer.__typename !== "User") throw notFound();
+  },
+
+  component: ChangePassword,
+});
+
+function ChangePassword(): React.ReactNode {
+  const { t } = useTranslation();
+  const [viewer] = useQuery({ query: CURRENT_VIEWER_QUERY });
+  if (viewer.error) throw viewer.error;
+  if (viewer.data?.viewer.__typename !== "User") throw notFound();
+  const userId = viewer.data.viewer.id;
+
+  const currentPasswordRef = useRef<HTMLInputElement>(null);
+  const newPasswordRef = useRef<HTMLInputElement>(null);
+  const newPasswordAgainRef = useRef<HTMLInputElement>(null);
+
+  const [result, changePassword] = useMutation(CHANGE_PASSWORD_MUTATION);
+
+  const onSubmit = (event: FormEvent<HTMLFormElement>): void => {
+    event.preventDefault();
+
+    const formData = new FormData(event.currentTarget);
+
+    const oldPassword = formData.get("current_password");
+    const newPassword = formData.get("new_password");
+    const newPasswordAgain = formData.get("new_password_again");
+
+    if (!oldPassword || !newPassword || !newPasswordAgain) {
+      throw new Error("not all passwords set");
+    }
+
+    // Assert that these are strings — according to the type system they could be Files
+    if (typeof oldPassword !== "string") {
+      throw new Error("oldPassword not string");
+    }
+    if (typeof newPassword !== "string") {
+      throw new Error("newPassword not string");
+    }
+
+    if (newPassword !== newPasswordAgain) {
+      throw new Error("passwords mismatch; this should be checked by the form");
+    }
+
+    changePassword({ userId, oldPassword, newPassword });
+  };
+
+  const success =
+    result.data && result.data.setPassword.status == SetPasswordStatus.Allowed;
+  const handleableError =
+    result.data && result.data.setPassword.status != SetPasswordStatus.Allowed;
+  const unhandleableError = result.error !== undefined;
+
+  const errorMsg: string | undefined = ((): string | undefined => {
+    switch (result.data?.setPassword.status) {
+      case SetPasswordStatus.InvalidNewPassword:
+        return t(
+          "frontend.password_change.failure.description_INVALID_NEW_PASSWORD",
+        );
+      case SetPasswordStatus.NoCurrentPassword:
+        return t(
+          "frontend.password_change.failure.description_NO_CURRENT_PASSWORD",
+        );
+      case SetPasswordStatus.NotAllowed:
+        return t("frontend.password_change.failure.description_NOT_ALLOWED");
+      case SetPasswordStatus.NotFound:
+        return t("frontend.password_change.failure.description_NOT_FOUND");
+      case SetPasswordStatus.PasswordChangesDisabled:
+        return t(
+          "frontend.password_change.failure.description_PASSWORD_CHANGES_DISABLED",
+        );
```

Comment:
Actually for those I would just `throw` within the component? None of those are not recoverable by the user, so showing a page 'crash' is probably the best thing?

Reply:
some of these I'm on the fence about.


e.g. password change disabled: the functionality could have been disabled since you navigated to the page, it feels a bit crap to show a generic 'something went wrong' error when we know the actual reason.

not allowed / not found: these probably make more sense if we had an admin control panel in MAS, but agreed we can probably lose them

no current password: I'd be tempted to keep it, since it's explanatory of a real-world error you might hit?